runtime/src/status_cache.rs

```diff
@@ -1,119 +1,159 @@
-                        if let Entry::Occupied(mut o_key_list) = all_hash_maps.entry(*key_slice) {
-                            let key_list = o_key_list.get_mut();
-                            key_list.retain(|(updated_slot, _)| *updated_slot != slot);
-                            if key_list.is_empty() {
-                                o_key_list.remove_entry();
-                            }
-                        } else {
-                            panic!(
-                                "Map for key must exist if key exists in self.slot_deltas, slot: {slot}"
-                            )
-                        }
-                    }
-
-                    if all_hash_maps.is_empty() {
-                        o_blockhash_entries.remove_entry();
-                    }
-                } else {
-                    panic!("Blockhash must exist if it exists in self.slot_deltas, slot: {slot}")
+    /// Clear all entries for a slot.
+    ///
+    /// This is used when a slot is purged from the bank, see
+    /// ReplayStage::purge_unconfirmed_duplicate_slot(). When this is called, it's guaranteed that
+    /// there are no threads inserting new entries for this slot, so there are no races.
+    pub fn clear_slot_entries(&self, slot: Slot) {
+        // remove txs seen during this slot
+        let Some((_, slot_deltas)) = self.slot_deltas.remove(&slot) else {
+            return;
+        };
+
+        // loop over all the blockhashes referenced by txs inserted in the slot
+        for item in slot_deltas.iter() {
+            // one of the blockhashes referenced
+            let blockhash = item.key();
+            // slot_delta_txs is self.slot_deltas[slot][blockhash], ie the txs that referenced
+            // `blockhash` in `slot`
+            let (_key_index, slot_delta_txs) = item.value();
+
+            // any self.slot_deltas[slot][blockhash] must also exist as self.cache[blockhash] - the
+            // two maps store the same blockhashes just in a different layout.
+            let Some(cache_guard) = self.cache.get(blockhash) else {
+                panic!("Blockhash must exist if it exists in self.slot_deltas, slot: {slot}")
+            };
+
+            // cache_txs is self.blockhash_cache[blockhash]
+            let (_, _, cache_txs) = &*cache_guard;
+
+            // loop over the txs in slot_delta[slot][blockhash]
+            for (_, (key_slice, _)) in slot_delta_txs {
+                // find the corresponding tx in self.cache[blockhash]
+                let Entry::Occupied(mut cache_tx_entry) = cache_txs.entry(*key_slice) else {
+                    panic!("Map for key must exist if key exists in self.slot_deltas, slot: {slot}")
+                };
+                // remove the slot from the list of forks this tx was inserted in
+                let forks = cache_tx_entry.get_mut();
+                forks.retain(|(fork_slot, _key)| *fork_slot != slot);
+                if forks.is_empty() {
```

Comment:
Might be over/prematurely optimizing... but this strikes me as a bit of an odd way to do this.

AFAIK, there is no chance of duplicate fork slots in the status cache. Seems we could check the len == 1, remove entry in that case. If != 1, then we can do retain.

Actually, retaining seems a bit odd to me because let's say we have 4 forks and we remove entry at 0. we shuffle everything over by 1. They're not **huge**, so copying over may be relatively cheap, but probably could just do something like this:

```rust
                // if all the slots have been cleared or purged, we don't need to track this tx
                // anymore
                if forks.len() == 1 {
                    cache_tx_entry.remove();
                } else {
                    forks
                        .iter()
                        .position(|(fork_slot, _)| *fork_slot == slot)
                        .map(|index| forks.swap_remove(index));
                }
```